### PR TITLE
🪛 Moving to SQLAlchemy 2.0

### PR DESCRIPTION
**Pull request by tarsil**
* Added support for python 3.11
* Added common and dialects packages to handle the new SQLAlchemy 2.0+

@aminalaee @Kludex this is a cleaner version of the previous PR with the changes passing the new SQLAlchemy 2.0 which is significantly different from the < 2.0

It is also good to mention that when I was doing this PR, I was reading some suggestions and the one from here https://github.com/encode/databases/pull/507 was good to add it as well. So I would like to thank @circlingthesun  and @JGoutin for the suggestions

Edit: The asyncpg sometimes hangs during the tests. I forcibly close the connections during that test to make sure it doesn't get stuck

### CODE REVIEW

**Comment by tarsil** [comment]
@aminalaee @Kludex maybe you could manually kill:

* https://github.com/encode/databases/actions/runs/4491971587
* https://github.com/encode/databases/actions/runs/4492090375

Please? This was the while reason why the second commit. Quite a well known issue but that CI passed with flying colours, only these 2 are hanging

**Comment by Kludex** [comment]
Done

**Comment by tarsil** [comment]
Well, I hope this is ok. Quite nice and clean integration for SQLAlchemy 2.0

**Comment by lucasguiss** [comment]
I've forked the version of this PR and it's working fine in our apps, waiting for the release 😄

**Comment by tarsil** [comment]
@lucasguiss very happy to hear that :)

**Comment by tarsil** [comment]
@Kludex here the status is not "complete" because this PR also makes sure python 3.7 is no longer supported for many reasons but also in June the support will stop.

**Comment by Kludex** [comment]
> @Kludex here the status is not "complete" because this PR also makes sure python 3.7 is no longer supported for many reasons but also in June the support will stop.

Why did you remove now?

**Comment by Kludex** [comment]
Can we remove all changes that are not related to the support of SQLAlchemy 2.0? You can create other PRs for them, but let's focus on a single thing here, please.

**Comment by tarsil** [comment]
> Can we remove all changes that are not related to the support of SQLAlchemy 2.0? You can create other PRs for them, but let's focus on a single thing here, please.

@Kludex Done. It was mainly that, had initial configuration issues with python 3.7 but that was addressed.

The changes you see are the ones used for the move to SQLAlchemy 2.0+

Regarding other PRs. I can do that later once the time comes. This was already a a good change that allow us to move forward.

File: CHANGELOG.md
Comment (Kludex) [inline CHANGELOG.md:1]: Please revert this.

File: .github/workflows/test-suite.yml
Comment (Kludex) [inline .github/workflows/test-suite.yml:1]: Please revert the formatting here. The 3.11 support is fine.

File: .github/workflows/publish.yml
Comment (Kludex) [inline .github/workflows/publish.yml:7]: ...

File: README.md
Comment (Kludex) [inline README.md:22]: Please open a new PR to remove 3.7.

File: README.md
Comment (Kludex) [inline README.md:18]: image

:thinking:

File: mkdocs.yml
Comment (Kludex) [inline mkdocs.yml:1]: Revert this please.

File: docs/index.md
Comment (Kludex) [inline docs/index.md:1]: Revert this please.

File: docs/index.md
Comment (Kludex) [inline docs/index.md:21]: You shouldn't use full path.

File: CHANGELOG.md
Comment (tarsil) [reply CHANGELOG.md:1]: Will do. This was automatic from my VSCode.

File: README.md
Comment (tarsil) [reply README.md:18]: Yap, got it. This were the docs. Reverting it. I can open another PR if that is the issue.

File: docs/index.md
Comment (tarsil) [reply docs/index.md:21]: This was my VSCode. Not even sure how it was done. Ok, let me have a look at everything

**Comment by tarsil** [comment]
Hey @aminalaee . Any news about this? I know you are extremely busy

**Comment by aminalaee** [comment]
Hi @tarsil , I will take a look today or tomorrow. Sorry for the delay.

**Comment by mathause** [comment]
I think you'll have to update `setup.py` as well:

https://github.com/encode/databases/blob/b6eba5f7a19eaf8966e3821f44fe00f4770cb822/setup.py#L50

**Comment by tarsil** [comment]
> I think you'll have to update `setup.py` as well:
> 
> https://github.com/encode/databases/blob/b6eba5f7a19eaf8966e3821f44fe00f4770cb822/setup.py#L50

I did update it as well before. Its there 👍🏼

**Comment by mathause** [comment]
pip complained when I installed the branch, but I see it now. Not sure why I missed it before - sorry for the noise.

**Comment by tarsil** [comment]
@aminalaee sorry to a bother but, any news?

File: CHANGELOG.md
Comment (Kludex) [inline CHANGELOG.md:9]:
```suggestion
- Support Python 3.11.
- Support for SQLAlchemy 2.0.
```

File: tests/test_connection_options.py
Comment (Kludex) [inline tests/test_connection_options.py:9]: How was this working before?

File: CHANGELOG.md
Comment (Kludex) [inline CHANGELOG.md:10]: Can we do this change on another PR? So I can really focus on what is needed for SQLA v2

File: tests/test_connection_options.py
Comment (tarsil) [reply tests/test_connection_options.py:9]: This was coming from the import. Although this works with any version anyway. No worries, rolling back now

File: CHANGELOG.md
Comment (tarsil) [reply CHANGELOG.md:10]: Rolling back now

**Review by Kludex — COMMENTED** [review]
It would be faster if you explain the non-trivial changes, like why they are there :sweat_smile:

File: tests/test_databases.py
Comment (rafalp) [reply tests/test_databases.py:169]: Passing columns as list is deprecated in 2.0:
 https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-core-usage

**Comment by ansipunk** [comment]
Maybe it's worth to call the new version `databases==1.0.0` which will require `SQLAlchemy>=2`, since the latter is not compatible with `<2` and get rid of compatibility glue from `databases` 🤔

**Comment by BulinLiang** [comment]
I set a default value for the date and name when defining the data field, but when adding a record, the default value was not created. Moreover, setting the default value as a function body does not allow for real-time generation of default values when adding data.
Here is the code.
~~~python
import sqlalchemy as sa 
User = sa.Table(
    "users",
    metadata,
    sa.Column("id", sa.Integer, primary_key=True),
    sa.Column("name", sa.String(), default="no_name"),
    sa.Column("created_on", sa.DateTime(), default=datetime.now, server_default=sa.func.now(), index=True)
)

db = Database('postgresql+asyncpg://xxx:xxx@localhost/sqlmodel')

async def create():
    query = User.insert()
    await db.execute(query=query, values=dict(version=0.12))

await db.connect()
await create()
~~~
Hope you can help me, thanks

**Comment by rafalp** [comment]
@BulinLiang please don't use pull request discussions for issue reporting unless your issue is directly related to the contents of pull request you are commenting on.

**Comment by eastack** [comment]
Can't wait, I want to know when we can merge and release.

**Comment by Kludex** [comment]
Some comments were made, but I guess someone needs to do a proper review here.

**Comment by tarsil** [comment]
@Kludex this ia probably my fault. Since our last conversation unfortunately I didn't have enough time to address these comments. @rafalp in your code comment. I know SQLA stopped supporting that and that is the reason why I didn't change too much and I simple added the * as per normal python. 

I promise I will try to address all the comments soon. Apologies

**Comment by MAKMED1337** [comment]
Few days ago SQLAlchemy released 2.11 where they have changed `Row` class
Will it be supported ?

File: databases/backends/common/records.py
Comment (MAKMED1337) [inline databases/backends/common/records.py:74]: Because of this you can't unpack result
Example
```
class LastUsername(Base):
	__tablename__ = 'last_username'
	user_id = Column(BIGINT, primary_key=True)
	username = Column(TEXT, nullable=False)

r = await db.fetch_one(select(LastUsername))
user_id, username = r
assert user_id == 'user_id' #but it need to be at least int
```

**Comment by stonebig** [comment]
hum, any hope for this to be merge ?

**Comment by Kludex** [comment]
> hum, any hope for this to be merged ?

If comments are addressed or answered, we can continue here

**Comment by Feijo** [comment]
Honestly I just decided to give up on this library altogether and use the new async features built-in into SA2.

**Comment by tarsil** [comment]
Sorry, didn't have enough time to fully address this but I will soon

**Comment by philipbjorge** [comment]
@tarsil -- 
Thanks for putting this together -- 
We've put this in one of our projects and our test suite was 🟢.

We'll be kicking the tires more on this over the coming weeks and sharing any feedback or discoveries here.

([We did pin on SA 2.0.10 based on this comment](https://github.com/encode/databases/pull/540#issuecomment-1529056689))

**Comment by ziglef** [comment]
Any updates or roadmap for this PR?

**Comment by philipbjorge** [comment]
We just shipped this branch out to production and haven't noticed any problems yet.
We're a fairly low volume service, but overall feeling very good about this branch.
We're using this as a bridge while we port over to SA2.

**Comment by adnam** [comment]
I've tried out this branch by specifying the following in my requirements file:
```
databases @ git+https://github.com/tarsil/databases@feature/add_sqlalchemy_2
```
which resulted in pip-compile choosing `sqlalchemy==2.0.23` - the latest version as of Nov 2 '23.

As previously [mentioned here](https://github.com/encode/databases/pull/540#issuecomment-1529056689) the Row class has changes in v2.0.11 resulting in an error like this when performing a query:
```
AttributeError: type object 'Row' has no attribute '_default_key_style'
```
I think it would be worth updating this PR to account for the new Row class before merging, otherwise projects using `databases` will be pinned to an sqlalchemy version from April '23.

**Comment by tarsil** [comment]
> @ansipunk solved this here [tarsil#2](https://github.com/tarsil/databases/pull/2) Just need to get this merged and we'll be back to 100% tests passing. What else needs to be done for this PR to be merged?

It is merged and I do apologize for not receiving notifications and missing this.

**Comment by Kludex** [comment]
When @tarsil is happy, I'll merge it and make a release. Ping me.

**Comment by tarsil** [comment]
@Kludex i do believe I just need to fix the conflicts and we should be good to go

**Comment by Kludex** [comment]
I've invited you to encode @tarsil . Ask me on PM if you have questions about workflows or anything.

**Review by Kludex — APPROVED** [review]
Merge when happy @tarsil 👍

**Comment by tarsil** [comment]
@Kludex , I just merged after fixing the minor conflicts (normal as it was an old branch) and passed 👍🏼

**Comment by tarsil** [comment]
> I've invited you to encode @tarsil . Ask me on PM if you have questions about workflows or anything.

Will do but so far was just this simple merge conflict, I believe. Thank you for the invite.

**Pull request merged**
